spacy/ml/tb_framework.pyx

```diff
@@ -148,32 +148,74 @@ def init(
     # model = _lsuv_init(model)
     return model
 
-InWithoutActions = Tuple[List[Doc], TransitionSystem]
-InWithActions = Tuple[List[Doc], TransitionSystem, List[Ints1d]]
-InT = TypeVar("InT", InWithoutActions, InWithActions)
-
-def forward(model, docs_moves: InT, is_train: bool):
-    if len(docs_moves) == 2:
-        docs, moves = docs_moves
-        actions = None
-    else:
-        docs, moves, actions = docs_moves
+
+class TransitionModelInputs:
+    """
+    Input to transition model.
+    """
+
+    # dataclass annotation is not yet supported in Cython 0.29.x,
+    # so, we'll do something close to it.
+
+    actions: Optional[List[Ints1d]]
+    docs: List[Doc]
+    max_moves: int
+    moves: TransitionSystem
+    states: Optional[List[State]]
+
+    __slots__ = [
+        "actions",
+        "docs",
+        "max_moves",
+        "moves",
+        "states",
+    ]
+
+    def __init__(self, docs: List[Doc], moves: TransitionSystem,
+        actions: Optional[List[Ints1d]]=None, max_moves: int=0,
+        states: Optional[List[State]]=None):
+        """
+        actions (Optional[List[Ints1d]]): actions to apply for each Doc.
+        docs (List[Doc]): Docs to predict transition sequences for.
+        max_moves: (Optional[int]): the maximum number of moves to apply,
+            values less than 1 will apply moves to states until they are
+            final states.
```

Comment:
A couple of things:
- `max_moves` isn't an optional by virtue of accepting a `None`, so we should update the description.
- On a similar note, given that the semantics of a zero value and is the same as that of any negative value, we could make this more explicit by restricting the lower-bound to zero and raising an exception when we receive a negative value.

Reply:
Fixed `max_moves` description.

> On a similar note, given that the semantics of a zero value and is the same as that of any negative value, we could make this more explicit by restricting the lower-bound to zero and raising an exception when we receive a negative value.

The pre-refactor parser does not have this restriction, so adding this restriction now could break existing configurations and it doesn't give us much.